protocol.md

```diff
@@ -0,0 +1,338 @@
+* proof-of-work equal `0.002`
+* proof-of-work time equal to `10`
+* time-to-live equal `10`
+
+If you want to receive messages from a mobile client, you need to run Whisper node with PoW set to `0.002`. In case of `geth`, this option can be overridden with `-shh.pow=0.002` flag.
+
+TODO: provide an instruction how to start a Whisper node with proper configuration using geth.
+
+## Keys management
+
+Protocol requires a key pair for the following actions:
+* signing a message,
+* decrypting received messages which are encrypted with user's public key.
+
+The private key is stored in memory in Whisper service because the communication happens through JSON-RPC and we don't want to send private keys directly every time a new message is sent and what's more important some actions like receiving and decrypting messages happen in the background asynchronously.
+
+Whisper also keeps symmetric keys for public chats for the same reason of receiving and decrypting messages asynchronously in the background.
+
+Keys management for PFS is described in [Perfect forward secrecy section](#perfect-forward-secrecy-pfs).
+
+## Encryption algorithms
+
+TBD (link to Whisper spec where all algorithms used to encrypting stuff are mentioned)
+
+## Topic
+
+There are two types of Whisper topics the protocol uses:
+* static topic for private chats (also called _contact discovery topic_)
+* dynamic topic based on a chat name for public chats
+
+The static topic is always the same and its hex representation is `0xf8946aac`. In fact, _the contact discovery topic_ is calculated using a dynamic topic algorithm described below with a constant name `contact-discovery`.
+
+Having only one topic for all private chats has an advantage as it's very hard to figure out who talks to who. A drawback is that everyone receives everyone's messages but they can decrypt only these they have private keys for.
+
+A dynamic topic is derived from a string using the following algorithm:
+
+```
+var hash []byte
+
+hash = keccak256(name)
+
+# Whisper V6 specific
+var topic [4]byte
+
+topic_len = 4
+
+if len(hash) < topic_len {
+    topic_len = len(hash)
+}
+
+for i = 0; i < topic_len; i++ {
+    topic[i] = hash[i]
+}
+```
+
+## Message encryption
+
+Protocol distinguishes messages encrypted using asymmetric and symmetric encryptions.
+
+Symmetric keys are created using `shh_generateSymKeyFromPassword` Whisper JSON-RPC method which accepts one param, a string.
+
+Messages encrypted with asymmetric encryption should be encrypted using recipient's public key so that only the recipient can decrypt them.
+
+# Perfect Forward Secrecy (PFS)
+
+TODO: link to a separate document (currently in the PR).
```

Comment:
I think we want this inline

Reply:
Then I would add it after this is merged. It's already pretty long. I will put a note to inline PFS document here. It probably needs to be adjusted a bit anyway.

Reply:
Yeah you are probably right